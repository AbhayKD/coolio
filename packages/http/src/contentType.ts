export enum ContentType {
  TEXT = 'text/plain',
  JSON = 'application/json',
  VND_JSON = 'application/vnd.api+json',
  MULTIPART_FORM = 'multipart/form-data',
  URL_ENCODED = 'application/x-www-form-urlencoded',
}

export const ContentTypeRegex = {
  TEXT: /^text\//,
  JSON: /^application\/(json|.+\+json)$/,
  URL_ENCODED: /^application\/x-www-form-urlencoded$/,
  MULTIPART: /^multipart\//,
};

export type ContentTypeMap<T> = Record<keyof typeof ContentTypeRegex, T>;

export function switchContentType<T>(contentType: string, map: ContentTypeMap<T>, defaultResult: T): T;
export function switchContentType<T>(contentType: string, map: ContentTypeMap<T>): T | undefined;
export function switchContentType<T>(
  contentType: string,
  map: ContentTypeMap<T>,
  defaultResult?: T,
) {
  for (const key in map) {
    if (map.hasOwnProperty(key)) {
      if (ContentTypeRegex[key].test(contentType)) {
        return map[key];
      }
    }
  }
  return defaultResult;
}
