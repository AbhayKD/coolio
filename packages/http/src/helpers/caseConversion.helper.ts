import { isArray, isPlainObject } from 'lodash';

export const splitWords = (text: string): string[] => {
  return text.split(/(?:[ _-]+)|(?=[A-Z]+)/);
};

export const deepKeyMap = (object: any, mapper: (key: string) => string): any => {
  if (isArray(object)) {
    return object.map(value => deepKeyMap(value, mapper));
  }
  if (isPlainObject(object)) {
    const result = {};
    for (const key in object) {
      if (object.hasOwnProperty(key)) {
        result[mapper(key)] = deepKeyMap(object[key], mapper);
      }
    }
    return result;
  }
  return object;
};

export const toCamelCase = (object: any) => deepKeyMap(object, key => splitWords(key)
  .map((word, index) => index > 0
    ? word[0].toUpperCase() + word.slice(1).toLowerCase()
    : word.toLowerCase())
  .join(''));

export const toPascalCase = (object: any) => deepKeyMap(object, key => splitWords(key)
  .map(word => word[0].toUpperCase() + word.slice(1).toLowerCase())
  .join(''));

export const toSnakeCase = (object: any) => deepKeyMap(object, key => splitWords(key)
  .map(word => word.toLowerCase())
  .join('_'));

export const toKebabCase = (object: any) => deepKeyMap(object, key => splitWords(key)
  .map(word => word.toLowerCase())
  .join('-'));

export const noConversion = (object: any) => object;
