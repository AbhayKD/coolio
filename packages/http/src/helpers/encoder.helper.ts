import TypedArray = NodeJS.TypedArray;
import { HttpResponseHeaders } from '../httpResponseHeaders';
import { Buffer } from 'buffer';

export type BufferEncoding =
  | 'ascii'
  | 'utf8'
  | 'utf-8'
  | 'utf16le'
  | 'ucs2'
  | 'ucs-2'
  | 'base64'
  | 'latin1'
  | 'binary'
  | 'hex'
  ;

/*
 * TextEncoder is not available on every environment, hence in this utility
 * we use "buffer" module (the npm one, not Node.js).
 */

export const getEncodingFromHeaders = (headers: HttpResponseHeaders | undefined, fallback = 'utf8') => (
  headers?.get('content-type')
    ?.split(';')
    ?.find(element => element.trim().startsWith('charset'))
    ?.split('=')[1]
  ?? fallback
).trim().toLowerCase();

export const encodeArrayBuffer = (data: TypedArray | string | undefined = '', encoding: BufferEncoding = 'utf8') =>
  typeof data === 'string'
    ? new Uint8Array(Buffer.from(data, encoding as BufferEncoding))
    : new Uint8Array(Buffer.from(data));

export const encodeText = (buffer: ArrayBuffer, encoding = 'utf8') => {
  return Buffer.from(buffer).toString(encoding);
};
