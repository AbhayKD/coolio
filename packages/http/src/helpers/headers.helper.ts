import isNil from 'lodash/isNil';
import { HttpHeaders } from '../httpClient.types';

export const getHeader = (headers: HttpHeaders | undefined, header: string): string | undefined => {
  if (!headers) {
    return undefined;
  }
  const lowercaseHeader = header.toLowerCase();
  const foundKey = Object.keys(headers).find(key => key.toLowerCase() === lowercaseHeader);
  return foundKey ? String(headers[foundKey]) : undefined;
};

export const parseHeaders = (headers: Headers): Record<string, string> => {
  const result: Record<string, string> = {};
  headers.forEach((value, key) => result[key.toLowerCase()] = value);
  return result;
};

export const sanitizeHeaders = (...multipleHeaders: (Record<string, any> | undefined)[]): Record<string, string> => {
  const result: Record<string, string> = {};
  for (const headers of multipleHeaders) {
    if (!headers) {
      continue;
    }
    for (const key in headers) {
      if (headers.hasOwnProperty(key)) {
        const resultKey = key.toLowerCase();
        const value = headers[key];
        if (isNil(value)) {
          delete result[resultKey];
        } else {
          result[resultKey] = value.toString();
        }
      }
    }
  }
  return result;
};
