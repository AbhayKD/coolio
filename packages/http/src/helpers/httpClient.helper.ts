import get from 'lodash/get';
import isNil from 'lodash/isNil';
import { ContentType } from '../httpClient.types';

const defaultHeaders = (host: string = '') => ({
  'Accept': 'application/json,application/vnd.api+json',
  'Content-Type': ContentType.JSON,
});

const toUrlEncoded = (obj: object) => {
  const str = [];
  for (const p in obj) {
    if (obj.hasOwnProperty(p) && !isNil(obj[p])) {
      str.push(encodeURIComponent(p) + '=' + encodeURIComponent(obj[p]));
    }
  }
  return str.join('&');
};

const sanitizeHeaders = (headers: Record<string, any>): Record<string, string> => {
  for (const key in headers) {
    if (!headers.hasOwnProperty(key)) {
      continue;
    }
    if (isNil(headers[key])) {
      delete headers[key];
    } else {
      headers[key] = headers[key].toString();
    }
  }
  return headers;
};

const getHostname = (url: string): string => {
  let hostname = '';
  try {
    hostname = new URL(url).hostname;
  } catch (e) {
    hostname = get(new RegExp(/https?:\/\/[^/]+/).exec(url), '[0]', '');
  }
  return hostname;
};

export const HttpClientHelper = {
  defaultHeaders,
  toUrlEncoded,
  sanitizeHeaders,
  getHostname,
};
