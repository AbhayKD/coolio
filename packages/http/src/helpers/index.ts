export * from './httpClient.helper';
export * from './headers.helper';
export * from './caseConversion.helper';
export * from './urlEncoding.helper';
