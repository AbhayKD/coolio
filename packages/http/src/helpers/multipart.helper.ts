const createFormDataImpl = (object: object, form: FormData, namespace?: string): FormData => {
  for (const key in object) {
    if (!object.hasOwnProperty(key) || object[key] === undefined) {
      continue;
    }
    const value = object[key];
    const formKey = namespace ? `${namespace}[${key}]` : key;
    if (Array.isArray(value)) {
      value.forEach((element, index) => {
        const arrayKey = `${formKey}[${index}]`;
        if (typeof element === 'object') {
          createFormDataImpl(element, form, arrayKey);
        } else {
          form.append(arrayKey, element);
        }
      });
    } else if (value instanceof Date) {
      form.append(formKey, value.toISOString());
    } else if (typeof value === 'object' && !(value instanceof File)) {
      createFormDataImpl(value, form, formKey);
    } else {
      form.append(formKey, value);
    }
  }
  return form;
};

export const createFormData = (body: object) => isFormData(body) ? body : createFormDataImpl(body, new FormData());

export const isFormData = (object: any): object is FormData => object instanceof FormData;

export const processMultipartBody = (body: string, boundary: string) => {
  body.split(boundary);
  throw new Error('Multipart response parsing is not implemented yet.');
};

export const getBoundaryFromContentTypeHeader = (header: string[]) => {
  const boundary = header.find(element => element.startsWith('boundary='))?.replace('boundary=', '');
  if (!boundary) {
    throw new Error('Could not find multipart boundary in Content-Type header.');
  }
  return boundary;
};
