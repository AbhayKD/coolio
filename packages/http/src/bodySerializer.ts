import isNil from 'lodash/isNil';
import isString from 'lodash/isString';
import isBuffer from 'lodash/isBuffer';
import { BodySerializer, HttpOptions, NormalizedHttpBody } from './httpClient.types';
import { BodyCasing, getCaseConverter, getHeader } from './helpers';
import { urlEncode } from './helpers/urlEncoding.helper';
import { ContentTypeMap, switchContentType } from './contentType';
import { createFormData, isFormData } from './helpers/multipart.helper';

export interface BodySerializerOptions {
  bodyCasing?: BodyCasing;
}

export const bodySerializer = ({
  bodyCasing,
}: BodySerializerOptions = {}): BodySerializer => {
  const caseConverter = getCaseConverter(bodyCasing);

  const bodySerializers: ContentTypeMap<(body: any) => any> = {
    JSON: (body) => JSON.stringify(caseConverter(body)),
    URL_ENCODED: (body) => urlEncode(caseConverter(body)),
    MULTIPART: (body) => body instanceof FormData ? body : createFormData(body),
    TEXT: (body) => String(body),
  };

  return (options: HttpOptions): NormalizedHttpBody => {
    const contentTypeHeader = getHeader(options.headers, 'content-type') || '';
    const contentType = contentTypeHeader.split(';')
      .map(type => type.trim().toLowerCase());
    const body = options.body;

    if (isNil(body) || isString(body) || isBuffer(body) || isFormData(body)) {
      return body as NormalizedHttpBody;
    }
    if (typeof body === 'object') {
      for (const type of contentType) {
        const serializedBody = switchContentType(type, bodySerializers);
        if (serializedBody) {
          return serializedBody(body);
        }
      }
    }
    throw new Error(`Can not serialize request body. Content-Type "${contentTypeHeader}"`);
  };
};
