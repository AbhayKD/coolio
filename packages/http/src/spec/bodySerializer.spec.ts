/* eslint-disable quote-props */
import { ContentType } from '../contentType';
import { bodySerializer } from '../bodySerializer';
import { BodyCasing, isFormData } from '../helpers';

describe('bodySerializer', () => {
  const jsonBodySample = Object.freeze({
    body: {
      test: {
        nested: [1, 2, 3],
      },
      casing: {
        'snake_case': 1,
        'camelCase': 2,
        'PascalCase': 3,
        'kebab-case': 4,
      },
      a: undefined,
      b: 'test',
      c: 0,
      d: false,
      e: null,
      f: '',
    }
  });

  const formDataBodySample = new FormData();
  formDataBodySample.append('abc', 'def');
  formDataBodySample.append('file', new Blob([
    Buffer.from('Zażółć gęślą jaźń.', 'utf8'),
  ]));

  it.each([
      [ContentType.JSON, BodyCasing.CAMEL_CASE],
      [ContentType.JSON, BodyCasing.KEBAB_CASE],
      [ContentType.JSON, BodyCasing.PASCAL_CASE],
      [ContentType.JSON, BodyCasing.SNAKE_CASE],
      [ContentType.URL_ENCODED, BodyCasing.CAMEL_CASE],
      [ContentType.URL_ENCODED, BodyCasing.KEBAB_CASE],
      [ContentType.URL_ENCODED, BodyCasing.PASCAL_CASE],
      [ContentType.URL_ENCODED, BodyCasing.SNAKE_CASE],
      [ContentType.MULTIPART_FORM, BodyCasing.CAMEL_CASE],
      [ContentType.MULTIPART_FORM, BodyCasing.KEBAB_CASE],
      [ContentType.MULTIPART_FORM, BodyCasing.PASCAL_CASE],
      [ContentType.MULTIPART_FORM, BodyCasing.SNAKE_CASE],
    ] as ([ContentType, BodyCasing])[]
  )('serializes request with %s body and %s casing', (contentType, bodyCasing) => {
    const body = bodySerializer({
      bodyCasing: bodyCasing,
    })({
      body: jsonBodySample,
      headers: {
        'Content-Type': contentType,
      }
    });
    expect(body).toMatchSnapshot(`${contentType}-with-${bodyCasing}`);
  });

  it('serializes request with text body', () => {
    const body = bodySerializer()({
      body: ['this is a text body'],
      headers: {
        'Content-Type': ContentType.TEXT,
      }
    });
    expect(body).toEqual('this is a text body');
  });

  it.each(
    [
      [undefined],
      [Buffer.from('buffer', 'utf8')],
      ['string'],
      [formDataBodySample]
    ] as [undefined | Buffer | string | FormData][]
  )('is a passthrough when %s is passed', (inputBody) => {
    const body = bodySerializer()({
      body: inputBody,
      headers: {},
    });
    expect(body).toBe(inputBody);
  });

  it('fails when Content-Type is not specified and non-passthrough item is passed', () => {
    expect(() => bodySerializer()({
      body: {},
      headers: {},
    })).toThrowError('Can not serialize request body');
  });

});
