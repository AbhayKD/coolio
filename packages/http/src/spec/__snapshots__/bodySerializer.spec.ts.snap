// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`bodySerializer serializes request with application/json body and CAMEL_CASE casing: application/json-with-CAMEL_CASE 1`] = `"{\\"body\\":{\\"test\\":{\\"nested\\":[1,2,3]},\\"casing\\":{\\"snakeCase\\":1,\\"camelCase\\":2,\\"pascalCase\\":3,\\"kebabCase\\":4},\\"b\\":\\"test\\",\\"c\\":0,\\"d\\":false,\\"e\\":null,\\"f\\":\\"\\"}}"`;

exports[`bodySerializer serializes request with application/json body and KEBAB_CASE casing: application/json-with-KEBAB_CASE 1`] = `"{\\"body\\":{\\"test\\":{\\"nested\\":[1,2,3]},\\"casing\\":{\\"snake-case\\":1,\\"camel-case\\":2,\\"pascal-case\\":3,\\"kebab-case\\":4},\\"b\\":\\"test\\",\\"c\\":0,\\"d\\":false,\\"e\\":null,\\"f\\":\\"\\"}}"`;

exports[`bodySerializer serializes request with application/json body and PASCAL_CASE casing: application/json-with-PASCAL_CASE 1`] = `"{\\"Body\\":{\\"Test\\":{\\"Nested\\":[1,2,3]},\\"Casing\\":{\\"SnakeCase\\":1,\\"CamelCase\\":2,\\"PascalCase\\":3,\\"KebabCase\\":4},\\"B\\":\\"test\\",\\"C\\":0,\\"D\\":false,\\"E\\":null,\\"F\\":\\"\\"}}"`;

exports[`bodySerializer serializes request with application/json body and SNAKE_CASE casing: application/json-with-SNAKE_CASE 1`] = `"{\\"body\\":{\\"test\\":{\\"nested\\":[1,2,3]},\\"casing\\":{\\"snake_case\\":1,\\"camel_case\\":2,\\"pascal_case\\":3,\\"kebab_case\\":4},\\"b\\":\\"test\\",\\"c\\":0,\\"d\\":false,\\"e\\":null,\\"f\\":\\"\\"}}"`;

exports[`bodySerializer serializes request with application/x-www-form-urlencoded body and CAMEL_CASE casing: application/x-www-form-urlencoded-with-CAMEL_CASE 1`] = `"body[test][nested][0]=1&body[test][nested][1]=2&body[test][nested][2]=3&body[casing][snakeCase]=1&body[casing][camelCase]=2&body[casing][pascalCase]=3&body[casing][kebabCase]=4&body[b]=test&body[c]=0&body[d]=false&body[e]=&body[f]="`;

exports[`bodySerializer serializes request with application/x-www-form-urlencoded body and KEBAB_CASE casing: application/x-www-form-urlencoded-with-KEBAB_CASE 1`] = `"body[test][nested][0]=1&body[test][nested][1]=2&body[test][nested][2]=3&body[casing][snake-case]=1&body[casing][camel-case]=2&body[casing][pascal-case]=3&body[casing][kebab-case]=4&body[b]=test&body[c]=0&body[d]=false&body[e]=&body[f]="`;

exports[`bodySerializer serializes request with application/x-www-form-urlencoded body and PASCAL_CASE casing: application/x-www-form-urlencoded-with-PASCAL_CASE 1`] = `"Body[Test][Nested][0]=1&Body[Test][Nested][1]=2&Body[Test][Nested][2]=3&Body[Casing][SnakeCase]=1&Body[Casing][CamelCase]=2&Body[Casing][PascalCase]=3&Body[Casing][KebabCase]=4&Body[B]=test&Body[C]=0&Body[D]=false&Body[E]=&Body[F]="`;

exports[`bodySerializer serializes request with application/x-www-form-urlencoded body and SNAKE_CASE casing: application/x-www-form-urlencoded-with-SNAKE_CASE 1`] = `"body[test][nested][0]=1&body[test][nested][1]=2&body[test][nested][2]=3&body[casing][snake_case]=1&body[casing][camel_case]=2&body[casing][pascal_case]=3&body[casing][kebab_case]=4&body[b]=test&body[c]=0&body[d]=false&body[e]=&body[f]="`;

exports[`bodySerializer serializes request with multipart/form-data body and CAMEL_CASE casing: multipart/form-data-with-CAMEL_CASE 1`] = `"FormData { body[test][nested][0]=1; body[test][nested][1]=2; body[test][nested][2]=3; body[casing][snakeCase]=1; body[casing][camelCase]=2; body[casing][pascalCase]=3; body[casing][kebabCase]=4; body[b]=test; body[c]=0; body[d]=false; body[f]=undefined }"`;

exports[`bodySerializer serializes request with multipart/form-data body and KEBAB_CASE casing: multipart/form-data-with-KEBAB_CASE 1`] = `"FormData { body[test][nested][0]=1; body[test][nested][1]=2; body[test][nested][2]=3; body[casing][snake-case]=1; body[casing][camel-case]=2; body[casing][pascal-case]=3; body[casing][kebab-case]=4; body[b]=test; body[c]=0; body[d]=false; body[f]=undefined }"`;

exports[`bodySerializer serializes request with multipart/form-data body and PASCAL_CASE casing: multipart/form-data-with-PASCAL_CASE 1`] = `"FormData { Body[Test][Nested][0]=1; Body[Test][Nested][1]=2; Body[Test][Nested][2]=3; Body[Casing][SnakeCase]=1; Body[Casing][CamelCase]=2; Body[Casing][PascalCase]=3; Body[Casing][KebabCase]=4; Body[B]=test; Body[C]=0; Body[D]=false; Body[F]=undefined }"`;

exports[`bodySerializer serializes request with multipart/form-data body and SNAKE_CASE casing: multipart/form-data-with-SNAKE_CASE 1`] = `"FormData { body[test][nested][0]=1; body[test][nested][1]=2; body[test][nested][2]=3; body[casing][snake_case]=1; body[casing][camel_case]=2; body[casing][pascal_case]=3; body[casing][kebab_case]=4; body[b]=test; body[c]=0; body[d]=false; body[f]=undefined }"`;
